### PR TITLE
Fix S3RepositoryThirdPartyTests.testReadFromPositionLargerThanBlobLength

### LINKED ISSUE #106457: [CI] S3RepositoryThirdPartyTests testReadFromPositionLargerThanBlobLength failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/2xqcn4leejnzi/tests/:modules:repository-s3:s3ThirdPartyTest/org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests/testReadFromPositionLargerThanBlobLength

**Reproduction line:**
```
./gradlew ':modules:repository-s3:s3ThirdPartyTest' --tests "org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests.testReadFromPositionLargerThanBlobLength" -Dtests.seed=5C8C88FF8345A464 -Dtests.locale=id -Dtests.timezone=Etc/GMT-9 -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
[Failure dashboard for `org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests#testReadFromPositionLargerThanBlobLength`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testReadFromPositionLargerThanBlobLength'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.Exception: Test abandoned because suite timeout was reached.

  at __randomizedtesting.SeedInfo.seed([5C8C88FF8345A464]:0)

```

### PR DESCRIPTION
**Pull request by tlrx**
The test should use a random operation purpose that is not "Indices", otherwise S3RetryingInputStream retries up to Integer.MAX_VALUE times which causes the test suite to timeout.

Also fixes the progress in the retries log messages.

Closes #106457

### CODE REVIEW

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java [hunk 1]
```diff
@@ -94,8 +94,8 @@ private void openStreamWithRetry() throws IOException {
                         : "requesting beyond end, start = " + start + " offset=" + currentOffset + " end=" + end;
                     getObjectRequest.setRange(Math.addExact(start, currentOffset), end);
                 }
-                final S3Object s3Object = SocketAccess.doPrivileged(() -> clientReference.client().getObject(getObjectRequest));
                 this.currentStreamFirstOffset = Math.addExact(start, currentOffset);
+                final S3Object s3Object = SocketAccess.doPrivileged(() -> clientReference.client().getObject(getObjectRequest));
                 this.currentStreamLastOffset = Math.addExact(currentStreamFirstOffset, getStreamLength(s3Object));
                 this.currentStream = s3Object.getObjectContent();
```
Comment (idegtiarenko) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:100]: Was it intentionally moved?

**Review by idegtiarenko — APPROVED** [review]
Thanks for a quick fix!

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java (hunk 1 above)
Comment (tlrx) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:100]: Yes. If the request fails, `currentStreamFirstOffset` is not updated and `currentStreamProgress()` reports an inaccurate number.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Hmm I think we shouldn't be retrying on a 416, regardless of the purpose. Could we try that instead.

**Comment by tlrx** [comment]
> I think we shouldn't be retrying on a 416, regardless of the purpose

I think you're right. I pushed some changes, let me know what you (both) think

**Review by idegtiarenko — APPROVED** [review]

File: modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3RepositoryThirdPartyTests.java [hunk 2]
```diff
@@ -239,9 +240,26 @@ public void testReadFromPositionLargerThanBlobLength() {
...
+                    + ", currentOffset=0] cannot be satisfied for blob object ["
+                    + repository.basePath().buildAsString()
+                    + blobName
+                    + ']'
+            )
+        );
+
+        var cause = exception.getRootCause();
+        assertThat(cause, instanceOf(AmazonS3Exception.class));
+        assertThat(((AmazonS3Exception) cause).getStatusCode(), equalTo(RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()));
```
Comment (DaveCTurner) [inline modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3RepositoryThirdPartyTests.java:263]: nit/one-liner suggestion:

```suggestion
        assertThat(
            asInstanceOf(AmazonS3Exception.class, exception.getRootCause()).getStatusCode(),
            equalTo(RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus())
        );
```

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3RepositoryThirdPartyTests.java (hunk 2 above)
Comment (tlrx) [reply modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3RepositoryThirdPartyTests.java]: TIL, I pushed 42ab562ca6097027876dbf8d90839d97d3d2567e

**Comment by tlrx** [comment]
Thanks both!

**Pull request merged**